pep-0xxx.rst

```diff
@@ -0,0 +1,571 @@
+the other classes defined here, these classes are not *abstract*, as they have
+no behaviour. They exist simply to signal certain common behaviours. Backends
+should subclass these exceptions in their own packages, but needn't define any
+behaviour for them. These exceptions should *never* be thrown directly, they
+should always be subclassed.
+
+The definitions of the errors are below::
+
+    class TLSError(Exception):
+        """
+        The base exception for all TLS related errors from any backend.
+        Catching this error should be sufficient to catch *all* TLS errors,
+        regardless of what backend is used.
+        """
+        pass
+
+    class WantWriteError(TLSError):
+        """
+        A special signaling exception used only when non-blocking or
+        buffer-only I/O is used. This error signals that the requested
+        operation cannot complete until more data is written to the network,
+        or until the output buffer is drained.
+        """
+        pass
+
+    class WantReadError(TLSError):
+        """
+        A special signaling exception used only when non-blocking or
+        buffer-only I/O is used. This error signals that the requested
+        operation cannot complete until more data is read from the network, or
+        until more data is available in the input buffer.
+        """
+        pass
+
+Changes to the Standard Library
+===============================
+
+The portions of the standard library that interact with TLS should be revised
+to use these ABCs. This will allow them to function with other TLS backends.
+This includes the following modules:
+
+- asyncio
+- ftplib
+- http.client
+- imaplib
+- nntplib
+- poplib
+- smtplib
+
+
+Future
+======
+
+Major future TLS features may require revisions of these ABCs. These revisions
+should be made cautiously: many backends may not be able to move forward
+swiftly, and will be invalidated by changes in these ABCs. This is acceptable,
+but wherever possible features that are specific to individual implementations
+should not be added to the ABCs. The ABCs should restrict themselves to
+high-level descriptions of IETF-specified features.
+
+
+ToDo
+====
+
+* Consider adding a new parameter (``valid_subjects``?) to ``wrap_socket`` and
+  ``wrap_buffers`` that specifies in a *typed* manner what kind of entries in
+  the SAN field are acceptable. This would break the union between SNI and
+  cert validation, which may be a good thing (you can't SNI an IP address, but
+  you can validate a cert with one if you want).
+* It's annoying that there's no type signature for fileobj. Do I really have to
+  define one as part of this PEP? Otherwise, how do I define the types of the
+  arguments to ``wrap_buffers``?
+* Do we need ways to disable specific TLS versions?
+* Do we need ways to control hostname validation?
+* Do we need ways to disable certificate validation altogether?
```

Comment:
Ok. To do this I have to define some level of optionality on this feature, because some TLS stacks may not be able to have their validation disabled (at least not easily).